src/sound/DalSoRiR2.cc

```diff
@@ -150,22 +150,48 @@ void DalSoRiR2::writeIO(word port, byte value, EmuTime::param time)
 	ymf278b.writeIO(port, value, time);
 }
 
-byte DalSoRiR2::readMem(word addr, EmuTime::param /*time*/)
+std::optional<byte> DalSoRiR2::readCommon(word addr) const
 {
+	auto read = [&](byte frame) {
+		return sram[(addr & 0x0FFF) + (frame & REG_FRAME_SEGMENT_MASK) * 0x1000];
+	};
+
 	if ((0x3000 <= addr) && (addr < 0x4000)) {
 		// SRAM frame 0
-		return sram[addr & 0x0FFF];
+		return read(0);
 	} else if ((0x7000 <= addr) && (addr < 0x8000) && ((regFrame[0] & DISABLE) == 0)) {
 		// SRAM frame 1
-		return sram[(addr & 0x0FFF) + (regFrame[0] & REG_FRAME_SEGMENT_MASK) * 0x1000];
+		return read(regFrame[0]);
 	} else if ((0xB000 <= addr) && (addr < 0xC000) && ((regFrame[1] & DISABLE) == 0)) {
 		// SRAM frame 2
-		return sram[(addr & 0x0FFF) + (regFrame[1] & REG_FRAME_SEGMENT_MASK) * 0x1000];
+		return read(regFrame[1]);
+	} else {
+		return {};
+	}
+}
+
+byte DalSoRiR2::readMem(word addr, EmuTime::param /*time*/)
+{
+	if (auto r = readCommon(addr)) {
+		return *r;
 	} else if (!dipSwitchBDIS.getBoolean()) {
-		// read selected segment of flash
-		auto flashBank = addr >> 14;
-		auto flashSegmentAddr = addr & 0x3FFF;
-		return flash.read(flashSegmentAddr + (regBank[flashBank] & REG_BANK_SEGMENT_MASK) * 0x4000);
+		auto page = addr >> 14;
+		auto bank = regBank[page] & REG_BANK_SEGMENT_MASK;
+		auto offset = addr & 0x3FFF;
+		return flash.read(0x4000 * bank + offset);
```

Comment:
You changed terminology here. I tried to stick to the manual's terminology... what's best? Use that, or try to be consistent across the openMSX codebase?

Reply:
I don't immediately see the conflict in terminology.
* The Z80 64kB address space is divided into exactly 4 pages of 16kB.
* (I'm skipping the MSX primary/secondary slot stuff).
* In each of the 4 pages it's possible to select a specific bank. For DalSoRi2 there are 32 possible banks.
* We then combine the selected bank and the offset with that bank (or page) to calculate the full address in the flash rom.

Or if the DalSoRiR2 uses 'page' and/or 'bank' with a different meaning, then personally I think we should stick to the openMSX terminology.

Reply:
In the variables I used in the original patch I directly copied the terminology from that manual: 
![image](https://github.com/user-attachments/assets/8fe7f554-1140-4a61-b905-9a1501a29270)
and
![image](https://github.com/user-attachments/assets/75a117d5-ebb2-453f-ac4c-10f5016f0c1b)

But, no deal breaker for me of course.